# Results:

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

Table of content : 
1. Global performances
2. Time
3. Top1 Count
4. Per Dataset analysis
5. Per architecture analysis

In [2]:
# path is different from output to circumvent the gitignore instructions
output_path = "results/"
save_repository = "results/"

In [3]:
# We also ran an experiment on a smaller set of datasets, using bagging over 10 weak learners for each method
use_bagging_results = 0 # 0 means no bagging, 1 means using bagging
suffix_save_path = "bagging_" * use_bagging_results

#### We ran 8060 experiments in regression, corresponding to 26 datasets for 31 methods with 10 random seeds.  In global average RMSE; AdaCap+SNN is first, GLU is second, CAT is fifth. 

In [7]:
if use_bagging_results:
    regression_df_formated = pd.read_csv(output_path+"regression_bagging.csv", index_col = 0)
    # New method name = method name + bagging 
    regression_df_formated["method"] = [value + "_bagging"*isbag for value, isbag in zip(regression_df_formated["method"].values,regression_df_formated["Use_bagging"].values)]
    # keep only first method seed to evaluate single learners
    regression_df_formated = regression_df_formated[regression_df_formated["method_seed"]==0]
else:
    regression_df_formated = pd.read_csv(output_path+"regression_benchmark.csv", index_col = 0)
regression_df_formated["R2"] = regression_df_formated["R2"].values.astype(float)
regression_df_formated["RMSE"] = np.sqrt(1. - regression_df_formated["R2"].values)

# 1. Global stats across all datasets averaged over 10 seeds


### 1.A Global stats: R2 score
In global average R2; CAT is first, AdaCap+SNN is second, GLU is third.

In [8]:
global_r2_order = regression_df_formated.groupby("method").mean().sort_values("R2").iloc[::-1].index.values
regression_df_formated[['dataset_seed', 'dataset_name',"method","R2"]].set_index("method").pivot_table(values="R2", index=['dataset_seed', 'dataset_name'],
                    columns=['method'], aggfunc=np.sum)[global_r2_order].describe().loc[["mean","std","min","25%","50%","75%","max"]]

method       CAT  adacapnetselu  adacapnetresblock  regularnetglu  \
mean    0.772448       0.771019           0.766774       0.766223   
std     0.223197       0.228893           0.226530       0.232569   
min    -0.402508      -0.267542          -0.100717      -0.391394   
25%     0.664486       0.621369           0.633756       0.631819   
50%     0.822513       0.849194           0.839521       0.854901   
75%     0.948972       0.958954           0.951684       0.952774   
max     0.996929       0.999133           0.999264       0.999099   

method  regularnetstandard  adacapnetbatchresblock  adacapnetstandard  \
mean              0.765300                0.764078           0.763541   
std               0.239274                0.224077           0.240056   
min              -0.698491               -0.100717          -0.264898   
25%               0.642985                0.635799           0.630829   
50%               0.849839                0.828021           0.839951   
75%               0.950078                0.954102           0.948954   
max               0.999191                0.999524           0.999198   

method  adacapnetfastselu  adacapnetfast  adacapnetglu  ...     NuSVM  \
mean             0.763255       0.761453      0.761330  ...  0.728550   
std              0.233555       0.231419      0.252776  ...  0.220941   
min             -0.356381      -0.228752     -0.386675  ... -0.174411   
25%              0.614657       0.625586      0.632503  ...  0.598709   
50%              0.833515       0.828784      0.856584  ...  0.790562   
75%              0.959755       0.945448      0.953509  ...  0.901779   
max              0.999572       0.998602      0.999556  ...  0.997028   

method       XRF      Enet     Lasso  regularnetbatchstandard      MARS  \
mean    0.720501  0.702501  0.702078                 0.701647  0.697824   
std     0.361331  0.248037  0.249883                 0.216932  0.295101   
min    -1.651110 -0.360848 -0.441897                -0.211250 -1.096739   
25%     0.607137  0.562069  0.563544                 0.586803  0.584504   
50%     0.821593  0.777346  0.777697                 0.728246  0.777522   
75%     0.922504  0.888873  0.889988                 0.865729  0.894973   
max     0.995957  0.986195  0.986017                 0.978427  0.984711   

method     Ridge      CART    Kernel     Intercept  
mean    0.661025  0.570802  0.542907 -1.576047e-02  
std     0.374696  0.461553  0.916477  4.387568e-02  
min    -1.792123 -2.800023 -6.440563 -4.926711e-01  
25%     0.542258  0.399216  0.522880 -1.141847e-02  
50%     0.760216  0.694567  0.754622 -1.557771e-03  
75%     0.891403  0.837555  0.893591 -1.844964e-04  
max     0.985860  0.985416  0.985887 -1.541660e-07  

[7 rows x 31 columns]

### 1.B Global stats: RMSE score

In [9]:
global_r2_order = regression_df_formated.groupby("method").mean().sort_values("RMSE").index.values
regression_df_formated[['dataset_seed', 'dataset_name',"method","RMSE"]].set_index("method").pivot_table(values="RMSE", index=['dataset_seed', 'dataset_name'],
                    columns=['method'], aggfunc=np.sum)[global_r2_order].describe().loc[["mean","std","min","25%","50%","75%","max"]]

method  adacapnetselu  regularnetglu  adacapnetglu  adacapnetresblock  \
mean         0.414782       0.420134      0.420630           0.421447   
std          0.239075       0.239761      0.248954           0.236268   
min          0.029453       0.030016      0.021080           0.027131   
25%          0.202551       0.217291      0.215582           0.219807   
50%          0.388337       0.380917      0.378703           0.400598   
75%          0.615330       0.606779      0.606215           0.605181   
max          1.125852       1.179574      1.177572           1.049151   

method       CAT  regularnetstandard  adacapnetstandard  adacapnetfastselu  \
mean    0.422151            0.423022           0.423317           0.424583   
std     0.222555            0.236575           0.239755           0.238101   
min     0.055419            0.028440           0.028313           0.020691   
25%     0.225883            0.223372           0.225895           0.200578   
50%     0.421289            0.387505           0.400056           0.407997   
75%     0.579236            0.597507           0.607595           0.620759   
max     1.184275            1.303262           1.124677           1.164638   

method  adacapnetbatchresblock  regularnetselu  ...  xgb_sklearn     NuSVM  \
mean                  0.425708        0.426012  ...     0.456943  0.474233   
std                   0.234318        0.240294  ...     0.237744  0.216178   
min                   0.021808        0.027692  ...     0.055007  0.054514   
25%                   0.214189        0.204294  ...     0.295625  0.313356   
50%                   0.414703        0.406079  ...     0.426240  0.457644   
75%                   0.603490        0.625070  ...     0.632234  0.633475   
max                   1.049151        1.112255  ...     1.395383  1.083702   

method      MARS      Enet     Lasso  regularnetbatchstandard     Ridge  \
mean    0.495034  0.500131  0.500362                 0.508979  0.521812   
std     0.239454  0.218061  0.218502                 0.198608  0.258737   
min     0.123648  0.117495  0.118250                 0.146879  0.118912   
25%     0.324079  0.333355  0.331681                 0.366425  0.329540   
50%     0.471674  0.471862  0.471489                 0.521297  0.489661   
75%     0.644561  0.661764  0.660648                 0.642804  0.676565   
max     1.448012  1.166554  1.200790                 1.100568  1.670965   

method    Kernel      CART  Intercept  
mean    0.559008  0.583542   1.007642  
std     0.381001  0.298360   0.020477  
min     0.118800  0.120765   1.000000  
25%     0.326204  0.403044   1.000092  
50%     0.495354  0.552646   1.000779  
75%     0.690727  0.775101   1.005693  
max     2.727740  1.949365   1.221749  

[7 rows x 31 columns]

### 1.C RMSE ratio 
- RMSE measures an error. It makes more sense to compare RMSE values in relative terms. If one dataset 1 the RMSE of new method is 0.001 and the RMSE of old method is 0.002, new method cuts error in half, which is massive. Meanwhile if on dataset 2, the RMSE of new method is 0.101 and the RMSE of old method is 0.1, new method increases error by 1%, which is negligeable. Yet when averaging raw RMSE values accross the two datasets, both methods have the same average RMSE of 0.051, which would indicate that new method is not usefull.
- Note that when averaging over 10 seeds, the method which performs best can have an average "Ratio_Best_RMSE_seed" above 1, if it did not uniformly outperformed all others over all seeds.

Metric "Ratio_Best_RMSE_seed" compares RMSE of a method on a dataset for a seed to minimal RMSE on same dataset with same seed : 
$$r_{best}(method_i, seed) = R\!M\!S\!E(method_i, seed) / min_{method_j}R\!M\!S\!E(method_j, seed)$$

In [10]:
regression_df_formated.set_index(["dataset_name","dataset_seed"], inplace = True)
regression_df_formated["Best_RMSE_seed"] = regression_df_formated.reset_index().groupby(["dataset_name","dataset_seed"]).min()["RMSE"]
regression_df_formated.reset_index(inplace = True)
regression_df_formated["Ratio_Best_RMSE_seed"] = (regression_df_formated["RMSE"] / regression_df_formated["Best_RMSE_seed"])

$\overline{r}_{best}$ is the average of the ratio $r_{best}(method, seed)$ over the 10 seeds.

In [9]:
regression_df_formated[['dataset_seed', 'dataset_name',"method","Ratio_Best_RMSE_seed"]].set_index("method").pivot_table(values="Ratio_Best_RMSE_seed", index=['dataset_name','dataset_seed'],
                    columns=['method'], aggfunc=np.mean)[regression_df_formated.groupby('method').mean().sort_values("Ratio_Best_RMSE_seed").index.values].describe().loc[["mean","std","min","25%","50%","75%","max"]]

method  adacapnetselu  adacapnetglu  regularnetglu  regularnetselu  \
mean         1.139175      1.152071       1.165654        1.179055   
std          0.148134      0.203107       0.201748        0.177228   
min          1.000000      1.000000       1.000000        1.000000   
25%          1.036208      1.037692       1.044721        1.056265   
50%          1.098299      1.094647       1.092601        1.130920   
75%          1.186219      1.163815       1.205491        1.246737   
max          2.129441      2.505845       2.295633        2.116076   

method  adacapnetresblock  adacapnetfastselu  adacapnetstandard  \
mean             1.182268           1.186416           1.195867   
std              0.226820           0.184181           0.260941   
min              1.000000           1.000000           1.000000   
25%              1.064823           1.044513           1.068379   
50%              1.112487           1.131039           1.118322   
75%              1.203532           1.272601           1.208240   
max              2.742705           1.938783           2.780260   

method  adacapnetbatchresblock  regularnetstandard  adacapnetfast  ...  \
mean                  1.200271            1.220299       1.228210  ...   
std                   0.219490            0.365312       0.299203  ...   
min                   1.000000            1.000000       1.000000  ...   
25%                   1.067970            1.053135       1.076666  ...   
50%                   1.121335            1.107551       1.140767  ...   
75%                   1.281662            1.215019       1.270311  ...   
max                   2.422610            4.322058       3.228660  ...   

method      LGBM      MARS      NuSVM       Enet      Lasso      Ridge  \
mean    1.599467  1.688426   1.716447   1.834633   1.835863   1.863537   
std     1.315630  1.078132   1.645271   1.530073   1.532184   1.504229   
min     1.000000  1.000000   1.000000   1.000000   1.000000   1.000000   
25%     1.045110  1.078865   1.071141   1.076575   1.076541   1.122867   
50%     1.121756  1.284192   1.204499   1.265571   1.259510   1.326181   
75%     1.507820  1.899551   1.495622   1.912961   1.912842   1.919075   
max     8.539295  8.901962  12.126352  12.061967  12.114954  11.925094   

method     Kernel       CART  regularnetbatchstandard  Intercept  
mean     1.902652   2.007749                 2.343440   5.369018  
std      1.491512   1.594833                 3.599478   7.234634  
min      1.000000   1.002347                 1.000000   1.031421  
25%      1.126098   1.385358                 1.067128   1.803875  
50%      1.421718   1.488266                 1.183701   2.917815  
75%      2.108355   1.703126                 1.779089   6.404077  
max     11.938569  11.661189                30.555804  50.605580  

[7 rows x 31 columns]

Metric "Ratio_CAT_RMSE_seed" compares RMSE of a method on a dataset for a seed to CatBoost RMSE on same dataset with same seed. This provides a comparison to assess in relative terms :
$$r_{C\!A\!T}(method_i, seed) = R\!M\!S\!E(method_i, seed) / R\!M\!S\!E(C\!A\!T, seed)$$

In [12]:
regression_df_formated.set_index(["dataset_name","dataset_seed"], inplace = True)
regression_df_formated["CAT_RMSE_seed"] = regression_df_formated[regression_df_formated["method"] == "CAT"]["RMSE"]
regression_df_formated.reset_index(inplace = True)
regression_df_formated["Ratio_CAT_RMSE_seed"] = regression_df_formated["RMSE"] / regression_df_formated["CAT_RMSE_seed"]

$\overline{r}_{Cat}$ is the average of the ratio $r_{Cat}(method_i, seed)$ over the 10 seeds.

In [13]:
regression_df_formated[['dataset_seed', 'dataset_name',"method","Ratio_CAT_RMSE_seed"]].set_index("method").pivot_table(values="Ratio_CAT_RMSE_seed", index=['dataset_name','dataset_seed'],
                    columns=['method'], aggfunc=np.mean)[regression_df_formated.groupby('method').mean().sort_values("Ratio_CAT_RMSE_seed").index.values].describe().loc[["mean","std","min","25%","50%","75%","max"]]

method  adacapnetselu  regularnetglu  adacapnetglu  adacapnetresblock  CAT  \
mean         0.983867       0.991217      0.991892           0.999954  1.0   
std          0.258397       0.245082      0.266702           0.238916  0.0   
min          0.145729       0.187779      0.145356           0.134242  1.0   
25%          0.948626       0.942252      0.943057           0.958389  1.0   
50%          1.015272       1.027238      1.020433           1.019230  1.0   
75%          1.103280       1.107552      1.101933           1.100820  1.0   
max          1.677345       1.638570      1.770698           1.860777  1.0   

method  regularnetstandard  adacapnetstandard  regularnetselu  \
mean              1.003704           1.007508        1.010595   
std               0.212262           0.240388        0.260070   
min               0.193143           0.140874        0.137017   
25%               0.966055           0.960554        0.976444   
50%               1.018989           1.022500        1.043607   
75%               1.106490           1.099304        1.126986   
max               1.789125           1.803003        2.116076   

method  adacapnetbatchresblock  regularnetresblock  ...   CATfast     NuSVM  \
mean                  1.016476            1.018829  ...  1.139577  1.217576   
std                   0.239299            0.252130  ...  0.242351  0.409188   
min                   0.107902            0.278597  ...  0.688023  0.551066   
25%                   0.970584            0.964934  ...  0.997108  1.011812   
50%                   1.029649            1.029950  ...  1.047229  1.121760   
75%                   1.126504            1.102198  ...  1.291643  1.301138   
max                   1.746896            3.163954  ...  2.575090  4.075330   

method      MARS      Enet     Lasso     Ridge    Kernel      CART  \
mean    1.312620  1.365432  1.365949  1.396717  1.435758  1.459175   
std     0.510510  0.585903  0.586973  0.577642  0.588434  0.376277   
min     0.356957  0.453914  0.455638  0.442467  0.437531  0.823140   
25%     1.017878  1.002067  1.004111  1.041020  1.052809  1.285265   
50%     1.112944  1.119348  1.128461  1.213158  1.255861  1.365491   
75%     1.499615  1.572579  1.545959  1.556393  1.637336  1.513899   
max     4.451063  4.476598  4.519085  4.417699  4.414622  2.779147   

method  regularnetbatchstandard  Intercept  
mean                   1.467487   3.591573  
std                    1.030454   2.921886  
min                    0.853352   0.851708  
25%                    1.018590   1.733133  
50%                    1.115814   2.402214  
75%                    1.446477   4.434600  
max                   10.335458  18.386053  

[7 rows x 31 columns]

### 1.D Global Ranking All metrics sorted by RMSE

In [12]:
regression_df_formated.groupby('method').mean().sort_values("Ratio_Best_RMSE_seed")[["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed"]]

RMSE        R2  Ratio_Best_RMSE_seed  \
method                                                              
adacapnetselu            0.414782  0.771019              1.139175   
adacapnetglu             0.420630  0.761330              1.152071   
regularnetglu            0.420134  0.766223              1.165654   
regularnetselu           0.426012  0.760994              1.179055   
adacapnetresblock        0.421447  0.766774              1.182268   
adacapnetfastselu        0.424583  0.763255              1.186416   
adacapnetstandard        0.423317  0.763541              1.195867   
adacapnetbatchresblock   0.425708  0.764078              1.200271   
regularnetstandard       0.423022  0.765300              1.220299   
adacapnetfast            0.428157  0.761453              1.228210   
regularnetresblock       0.428298  0.754150              1.248691   
regularnetfastselu       0.439162  0.750726              1.256094   
regularnetfast           0.431209  0.761213              1.272808   
adacapnetbatchstandard   0.445570  0.747873              1.328215   
CAT                      0.422151  0.772448              1.390193   
XRF                      0.455160  0.720501              1.418276   
regularnetbatchresblock  0.443639  0.741890              1.437460   
xgb_sklearn              0.456943  0.734898              1.511431   
XGBoost                  0.452848  0.736771              1.513013   
RF                       0.442127  0.758483              1.540979   
CATfast                  0.455446  0.743761              1.544975   
LGBM                     0.454772  0.746796              1.599467   
MARS                     0.495034  0.697824              1.688426   
NuSVM                    0.474233  0.728550              1.716447   
Enet                     0.500131  0.702501              1.834633   
Lasso                    0.500362  0.702078              1.835863   
Ridge                    0.521812  0.661025              1.863537   
Kernel                   0.559008  0.542907              1.902652   
CART                     0.583542  0.570802              2.007749   
regularnetbatchstandard  0.508979  0.701647              2.343440   
Intercept                1.007642 -0.015760              5.369018   

                         Ratio_CAT_RMSE_seed  
method                                        
adacapnetselu                       0.983867  
adacapnetglu                        0.991892  
regularnetglu                       0.991217  
regularnetselu                      1.010595  
adacapnetresblock                   0.999954  
adacapnetfastselu                   1.023041  
adacapnetstandard                   1.007508  
adacapnetbatchresblock              1.016476  
regularnetstandard                  1.003704  
adacapnetfast                       1.024992  
regularnetresblock                  1.018829  
regularnetfastselu                  1.062783  
regularnetfast                      1.038209  
adacapnetbatchstandard              1.096827  
CAT                                 1.000000  
XRF                                 1.090935  
regularnetbatchresblock             1.119871  
xgb_sklearn                         1.124254  
XGBoost                             1.089023  
RF                                  1.110607  
CATfast                             1.139577  
LGBM                                1.139569  
MARS                                1.312620  
NuSVM                               1.217576  
Enet                                1.365432  
Lasso                               1.365949  
Ridge                               1.396717  
Kernel                              1.435758  
CART                                1.459175  
regularnetbatchstandard             1.467487  
Intercept                           3.591573

- Note that when averaging over 10 seeds, the method which performs best can have an average "Ratio_Best_RMSE_seed" above 1, if it did not uniformly outperformed all others over all seeds.

# 2. Time comparison of top k best methods
- Time in seconds averaged over all datasets

In [5]:
k = 10 #Up to 31 methods
regression_df_formated.groupby("method").mean().sort_values("RMSE")["time"].iloc[:k]

method
adacapnetselu              19.798833
regularnetglu               9.891143
adacapnetglu               22.355343
adacapnetresblock          17.192343
CAT                        92.518085
regularnetstandard          4.058149
adacapnetstandard          17.208227
adacapnetfastselu           7.667071
adacapnetbatchresblock    194.728147
regularnetselu              7.189570
Name: time, dtype: float64

# 3. Top 1 performance count

In [14]:
metric_id = 0 # Value between 0 and 4
metric = ["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed", "time"][metric_id]

granularity_id = 0 # Value between 0 and 1
granularity = ["method","category"][granularity_id]

metric_order = [1,-1,1,1,1][metric_id]
pd.DataFrame.from_dict(data = {**{key:0 for key in regression_df_formated[granularity].unique()},**{key:value for key, value in regression_df_formated.groupby(["dataset_name","method","category"]).mean().reset_index().sort_values(["dataset_name",metric]).iloc[::metric_order].groupby(["dataset_name"]).first()[[granularity,metric]].groupby(granularity).count().sort_values(metric).iloc[::metric_order][metric].iteritems()}}, orient= "index", columns = ["#Datasets with "+ metric +" Top1 performance"]).sort_values("#Datasets with "+ metric +" Top1 performance").iloc[::-1]

#Datasets with RMSE Top1 performance
CAT                                                         8
adacapnetglu                                                3
xgb_sklearn                                                 3
XRF                                                         2
Kernel                                                      1
adacapnetselu                                               1
adacapnetfastselu                                           1
CATfast                                                     1
Ridge                                                       1
regularnetfastselu                                          1
NuSVM                                                       1
regularnetglu                                               1
RF                                                          1
Enet                                                        1
adacapnetstandard                                           0
Lasso                                                       0
regularnetbatchresblock                                     0
XGBoost                                                     0
LGBM                                                        0
MARS                                                        0
regularnetresblock                                          0
regularnetbatchstandard                                     0
regularnetfast                                              0
regularnetselu                                              0
regularnetstandard                                          0
CART                                                        0
adacapnetbatchresblock                                      0
adacapnetbatchstandard                                      0
adacapnetfast                                               0
adacapnetresblock                                           0
Intercept                                                   0

# 4. Per Dataset Granularity

### 4.A Best performing method per dataset (averaged over 10 seeds)

In [15]:
regression_df_formated.groupby(["dataset_name","method","category"]).mean().reset_index().sort_values(["dataset_name","R2"]).groupby("dataset_name").last()[["category","method","n","p", "RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed"]]

category  \
dataset_name                                                  
Abaloneabalone.data_task_R_target_-1                     NN   
Appliances energy predictionenergydata_complete...       RF   
Beijing PM2.5 DataPRSA_data_2010.1.1-2014.12.31...       NN   
Communities and Crimecommunities.data_task_R_ta...      GLM   
Computer Hardwaremachine.data_task_R_target_-1       ADACAP   
Concrete Slump Testslump_test.data_task_R_targe...   ADACAP   
Concrete Slump Testslump_test.data_task_R_targe...   ADACAP   
Concrete Slump Testslump_test.data_task_R_targe...      SVM   
Electrical Grid Stability Simulated Data Data_f...   ADACAP   
Optical Interconnection Network optical_interco...     GBDT   
Optical Interconnection Network optical_interco...     GBDT   
Optical Interconnection Network optical_interco...     GBDT   
Physicochemical Properties of Protein Tertiary ...       RF   
QSAR Bioconcentration classes datasetGrisoni_et...     GBDT   
QSAR aquatic toxicityqsar_aquatic_toxicity.csv_...     GBDT   
QSAR fish toxicityqsar_fish_toxicity.csv_task_R...     GBDT   
Seoul Bike Sharing DemandSeoulBikeData.csv_task...     GBDT   
Servoservo.data_task_R_target_-1                       GBDT   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas...   ADACAP   
arashnic_marketing-seris-customer-lifetime-valu...     GBDT   
harlfoxem_housesalespredictionkc_house_data.csv...     GBDT   
how-much-did-it-rain-iitrain_task_R_target_-1          GBDT   
mercedes-benz-greener-manufacturingtrain.csv_ta...       RF   
mirichoi0218_insuranceinsurance.csv_task_R_targ...      SVM   
wiki4HEwiki4HE.csv_task_R_target_-10                    GLM   
wiki4HEwiki4HE.csv_task_R_target_-11                   GBDT   

                                                                method  \
dataset_name                                                             
Abaloneabalone.data_task_R_target_-1                regularnetfastselu   
Appliances energy predictionenergydata_complete...                 XRF   
Beijing PM2.5 DataPRSA_data_2010.1.1-2014.12.31...       regularnetglu   
Communities and Crimecommunities.data_task_R_ta...               Ridge   
Computer Hardwaremachine.data_task_R_target_-1           adacapnetselu   
Concrete Slump Testslump_test.data_task_R_targe...       adacapnetselu   
Concrete Slump Testslump_test.data_task_R_targe...   adacapnetfastselu   
Concrete Slump Testslump_test.data_task_R_targe...              Kernel   
Electrical Grid Stability Simulated Data Data_f...        adacapnetglu   
Optical Interconnection Network optical_interco...                 CAT   
Optical Interconnection Network optical_interco...                 CAT   
Optical Interconnection Network optical_interco...                 CAT   
Physicochemical Properties of Protein Tertiary ...                 XRF   
QSAR Bioconcentration classes datasetGrisoni_et...                 CAT   
QSAR aquatic toxicityqsar_aquatic_toxicity.csv_...                 CAT   
QSAR fish toxicityqsar_fish_toxicity.csv_task_R...                 CAT   
Seoul Bike Sharing DemandSeoulBikeData.csv_task...                 CAT   
Servoservo.data_task_R_target_-1                                   CAT   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas...        adacapnetglu   
arashnic_marketing-seris-customer-lifetime-valu...         xgb_sklearn   
harlfoxem_housesalespredictionkc_house_data.csv...             CATfast   
how-much-did-it-rain-iitrain_task_R_target_-1              xgb_sklearn   
mercedes-benz-greener-manufacturingtrain.csv_ta...                  RF   
mirichoi0218_insuranceinsurance.csv_task_R_targ...               NuSVM   
wiki4HEwiki4HE.csv_task_R_target_-10                              Enet   
wiki4HEwiki4HE.csv_task_R_target_-11                       xgb_sklearn   

                                                          n      p      RMSE  \
dataset_name                                                                   
Abaloneabalone.data_task_R_target_-1                 

### 4.B Metric score for each method for each dataset averaged over 10 seeds

In [6]:
metric_id = 0 # Value between 0 and 5
metric = ["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed", "time"][metric_id]
metric_order = [1,-1,1,1,-1][metric_id]
global_order = regression_df_formated.groupby("method").mean().sort_values(metric).iloc[::metric_order].index.values
RMSEs = regression_df_formated[['dataset_seed', 'dataset_name',"method",metric]].set_index("method").pivot_table(values=metric, index=['dataset_name'],
                    columns=['method'], aggfunc=np.mean)[global_order]
RMSEs

method                                              adacapnetselu  \
dataset_name                                                        
Abaloneabalone.data_task_R_target_-1                     0.651050   
Appliances energy predictionenergydata_complete...       0.602415   
Beijing PM2.5 DataPRSA_data_2010.1.1-2014.12.31...       0.397356   
Communities and Crimecommunities.data_task_R_ta...       0.573783   
Computer Hardwaremachine.data_task_R_target_-1           0.035146   
Concrete Slump Testslump_test.data_task_R_targe...       0.103122   
Concrete Slump Testslump_test.data_task_R_targe...       0.307871   
Concrete Slump Testslump_test.data_task_R_targe...       0.359230   
Electrical Grid Stability Simulated Data Data_f...       0.168163   
Optical Interconnection Network optical_interco...       0.243593   
Optical Interconnection Network optical_interco...       0.113960   
Optical Interconnection Network optical_interco...       0.246728   
Physicochemical Properties of Protein Tertiary ...       0.157761   
QSAR Bioconcentration classes datasetGrisoni_et...       0.473772   
QSAR aquatic toxicityqsar_aquatic_toxicity.csv_...       0.735627   
QSAR fish toxicityqsar_fish_toxicity.csv_task_R...       0.629031   
Seoul Bike Sharing DemandSeoulBikeData.csv_task...       0.293723   
Servoservo.data_task_R_target_-1                         0.344942   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas...       0.121072   
arashnic_marketing-seris-customer-lifetime-valu...       0.393560   
harlfoxem_housesalespredictionkc_house_data.csv...       0.679083   
how-much-did-it-rain-iitrain_task_R_target_-1            0.649060   
mercedes-benz-greener-manufacturingtrain.csv_ta...       0.316501   
mirichoi0218_insuranceinsurance.csv_task_R_targ...       0.987899   
wiki4HEwiki4HE.csv_task_R_target_-10                     0.585657   
wiki4HEwiki4HE.csv_task_R_target_-11                     0.614214   

method                                              regularnetglu  \
dataset_name                                                        
Abaloneabalone.data_task_R_target_-1                     0.657238   
Appliances energy predictionenergydata_complete...       0.586115   
Beijing PM2.5 DataPRSA_data_2010.1.1-2014.12.31...       0.346919   
Communities and Crimecommunities.data_task_R_ta...       0.575380   
Computer Hardwaremachine.data_task_R_target_-1           0.037094   
Concrete Slump Testslump_test.data_task_R_targe...       0.141798   
Concrete Slump Testslump_test.data_task_R_targe...       0.369433   
Concrete Slump Testslump_test.data_task_R_targe...       0.427332   
Electrical Grid Stability Simulated Data Data_f...       0.158926   
Optical Interconnection Network optical_interco...       0.223706   
Optical Interconnection Network optical_interco...       0.116831   
Optical Interconnection Network optical_interco...       0.269972   
Physicochemical Properties of Protein Tertiary ...       0.132245   
QSAR Bioconcentration classes datasetGrisoni_et...       0.479421   
QSAR aquatic toxicityqsar_aquatic_toxicity.csv_...       0.751963   
QSAR fish toxicityqsar_fish_toxicity.csv_task_R...       0.656061   
Seoul Bike Sharing DemandSeoulBikeData.csv_task...       0.290736   
Servoservo.data_task_R_target_-1                         0.346844   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas...       0.117870   
arashnic_marketing-seris-customer-lifetime-valu...       0.399690   
harlfoxem_housesalespredictionkc_house_data.csv...       0.678761   
how-much-did-it-rain-iitrain_task_R_target_-1            0.661681   
mercedes-benz-greener-manufacturingtrain.csv_ta...       0.333969   
mirichoi0218_insuranceinsurance.csv_task_R_targ...       1.002677   
wiki4HEwiki4HE.csv_task_R_target_-10                     0.566377   
wiki4HEwiki4HE.csv_task_R_target_-11                     0.594437   

method                                              adacapnetglu  \
dataset_name                                                      

In [8]:
# filtering columns to retain only adacap and regularnet
cols = RMSEs.columns
net_methods = []
for col in cols:
    if col.startswith("adacapnet"):
        net_methods.append(col.replace("adacapnet", ""))
new_columns = []
for method in net_methods:
    new_columns.append("adacapnet"+method)
    new_columns.append("regularnet"+method)

In [9]:
RMSEs[new_columns]

method                                              adacapnetselu  \
dataset_name                                                        
Abaloneabalone.data_task_R_target_-1                     0.651050   
Appliances energy predictionenergydata_complete...       0.602415   
Beijing PM2.5 DataPRSA_data_2010.1.1-2014.12.31...       0.397356   
Communities and Crimecommunities.data_task_R_ta...       0.573783   
Computer Hardwaremachine.data_task_R_target_-1           0.035146   
Concrete Slump Testslump_test.data_task_R_targe...       0.103122   
Concrete Slump Testslump_test.data_task_R_targe...       0.307871   
Concrete Slump Testslump_test.data_task_R_targe...       0.359230   
Electrical Grid Stability Simulated Data Data_f...       0.168163   
Optical Interconnection Network optical_interco...       0.243593   
Optical Interconnection Network optical_interco...       0.113960   
Optical Interconnection Network optical_interco...       0.246728   
Physicochemical Properties of Protein Tertiary ...       0.157761   
QSAR Bioconcentration classes datasetGrisoni_et...       0.473772   
QSAR aquatic toxicityqsar_aquatic_toxicity.csv_...       0.735627   
QSAR fish toxicityqsar_fish_toxicity.csv_task_R...       0.629031   
Seoul Bike Sharing DemandSeoulBikeData.csv_task...       0.293723   
Servoservo.data_task_R_target_-1                         0.344942   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas...       0.121072   
arashnic_marketing-seris-customer-lifetime-valu...       0.393560   
harlfoxem_housesalespredictionkc_house_data.csv...       0.679083   
how-much-did-it-rain-iitrain_task_R_target_-1            0.649060   
mercedes-benz-greener-manufacturingtrain.csv_ta...       0.316501   
mirichoi0218_insuranceinsurance.csv_task_R_targ...       0.987899   
wiki4HEwiki4HE.csv_task_R_target_-10                     0.585657   
wiki4HEwiki4HE.csv_task_R_target_-11                     0.614214   

method                                              regularnetselu  \
dataset_name                                                         
Abaloneabalone.data_task_R_target_-1                      0.654178   
Appliances energy predictionenergydata_complete...        0.627199   
Beijing PM2.5 DataPRSA_data_2010.1.1-2014.12.31...        0.493298   
Communities and Crimecommunities.data_task_R_ta...        0.577995   
Computer Hardwaremachine.data_task_R_target_-1            0.037332   
Concrete Slump Testslump_test.data_task_R_targe...        0.123295   
Concrete Slump Testslump_test.data_task_R_targe...        0.309741   
Concrete Slump Testslump_test.data_task_R_targe...        0.399237   
Electrical Grid Stability Simulated Data Data_f...        0.170603   
Optical Interconnection Network optical_interco...        0.245942   
Optical Interconnection Network optical_interco...        0.109288   
Optical Interconnection Network optical_interco...        0.254880   
Physicochemical Properties of Protein Tertiary ...        0.158993   
QSAR Bioconcentration classes datasetGrisoni_et...        0.472785   
QSAR aquatic toxicityqsar_aquatic_toxicity.csv_...        0.732276   
QSAR fish toxicityqsar_fish_toxicity.csv_task_R...        0.639697   
Seoul Bike Sharing DemandSeoulBikeData.csv_task...        0.317211   
Servoservo.data_task_R_target_-1                          0.362804   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas...        0.121051   
arashnic_marketing-seris-customer-lifetime-valu...        0.398643   
harlfoxem_housesalespredictionkc_house_data.csv...        0.679832   
how-much-did-it-rain-iitrain_task_R_target_-1             0.650739   
mercedes-benz-greener-manufacturingtrain.csv_ta...        0.322971   
mirichoi0218_insuranceinsurance.csv_task_R_targ...        1.009929   
wiki4HEwiki4HE.csv_task_R_target_-10                      0.588886   
wiki4HEwiki4HE.csv_task_R_target_-11                      0.617522   

method                                              adacapnetglu  \
dataset_name                          

In [17]:
for metric_id in range(5):
    metric = ["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed", "time"][metric_id]
    metric_order = [1,-1,1,1,-1][metric_id]
    global_order = regression_df_formated.groupby("method").mean().sort_values(metric).iloc[::metric_order].index.values
    regression_df_formated[['dataset_seed', 'dataset_name',"method",metric]].set_index("method").pivot_table(values=metric, index=['dataset_name'],
                        columns=['method'], aggfunc=np.mean)[global_order].to_csv(output_path+save_repository+"dataset_by_method_"+suffix_save_path+metric+".csv")

### 4.C Zoom on a specific dataset

In [18]:
dataset_index = 0 # between 0 and 25
dataset_name = regression_df_formated["dataset_name"].unique()[dataset_index]
print("DATASET : "+dataset_name)
table_mean = regression_df_formated.groupby(["dataset_name","method","category"]).mean().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']]
table_std = regression_df_formated.groupby(["dataset_name","method","category"]).std().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']]
table_min = regression_df_formated.groupby(["dataset_name","method","category"]).min().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']]
table_max = regression_df_formated.groupby(["dataset_name","method","category"]).max().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']]
for metric in ["R2", "RMSE"]:
    table_mean[metric + ":std"] = table_std[metric]
    table_mean[metric + ":min"] = table_min[metric]
    table_mean[metric + ":max"] = table_max[metric]
table_mean[['category']+[metric+suffix for metric in ["RMSE","R2"] for suffix in ["", ":std",":min", ":max"]]+ [ 'n', 'p', 'time']]

DATASET : mirichoi0218_insuranceinsurance.csv_task_R_target_-1


category      RMSE  RMSE:std  RMSE:min  RMSE:max  \
method                                                                      
NuSVM                         SVM  0.957468  0.068535  0.855221  1.083702   
adacapnetbatchresblock     ADACAP  0.977853  0.066826  0.813040  1.049151   
adacapnetresblock          ADACAP  0.980095  0.067166  0.813040  1.049151   
adacapnetselu              ADACAP  0.987899  0.082149  0.819413  1.125852   
adacapnetfast              ADACAP  0.991107  0.086218  0.834141  1.108491   
CATfast                      GBDT  0.991513  0.117019  0.838159  1.200091   
regularnetfast                 NN  0.993271  0.085683  0.829759  1.112328   
regularnetfastselu             NN  0.997177  0.064996  0.920170  1.148371   
regularnetbatchstandard        NN  0.997442  0.064485  0.882504  1.100568   
regularnetbatchresblock        NN  0.997497  0.103698  0.873633  1.205637   
CAT                          GBDT  0.998832  0.118787  0.844555  1.184275   
regularnetglu                  NN  1.002677  0.073788  0.933532  1.179574   
RF                             RF  1.007029  0.113476  0.883046  1.215964   
adacapnetfastselu          ADACAP  1.008141  0.081821  0.884006  1.164638   
regularnetselu                 NN  1.009929  0.061894  0.916555  1.112255   
adacapnetbatchstandard     ADACAP  1.015803  0.085081  0.864908  1.147966   
adacapnetstandard          ADACAP  1.017145  0.082787  0.864908  1.124677   
regularnetstandard             NN  1.017816  0.118279  0.883976  1.303262   
Enet                          GLM  1.021125  0.053821  0.984532  1.166554   
Intercept                Baseline  1.021134  0.021513  1.001097  1.056594   
Lasso                         GLM  1.027821  0.063624  0.982417  1.200790   
LGBM                         GBDT  1.032736  0.120610  0.892359  1.231731   
regularnetresblock             NN  1.041671  0.124313  0.877948  1.252758   
adacapnetglu               ADACAP  1.065250  0.055430  1.010076  1.177572   
xgb_sklearn                  GBDT  1.111538  0.174349  0.893083  1.395383   
XGBoost                      GBDT  1.123172  0.183653  0.953049  1.467391   
MARS                         MARS  1.133637  0.157580  0.978292  1.448012   
XRF                            RF  1.314921  0.189445  1.042588  1.628223   
Ridge                         GLM  1.350212  0.175539  1.115277  1.670965   
CART                         TREE  1.460823  0.244833  1.077524  1.949365   
Kernel                        SVM  2.115787  0.384301  1.666264  2.727740   

                               R2    R2:std    R2:min    R2:max      n      p  \
method                                                                          
NuSVM                    0.079027  0.133280 -0.174411  0.268597  109.0  330.0   
adacapnetbatchresblock   0.039784  0.124371 -0.100717  0.338966  109.0  330.0   
adacapnetresblock        0.035354  0.124966 -0.100717  0.338966  109.0  330.0   
adacapnetselu            0.017981  0.160006 -0.267542  0.328562  109.0  330.0   
adacapnetfast            0.011017  0.167162 -0.228752  0.304209  109.0  330.0   
CATfast                  0.004578  0.235758 -0.440217  0.297489  109.0  330.0   
regularnetfast           0.006805  0.166676 -0.237274  0.311500  109.0  330.0   
regularnetfastselu       0.001835  0.134553 -0.318755  0.153287  109.0  330.0   
regularnetbatchstandard  0.001366  0.128835 -0.211250  0.221187  109.0  330.0   
regularnetbatchresblock -0.004678  0.217799 -0.453561  0.236766  109.0  330.0   
CAT                     -0.010365  0.240327 -0.402508  0.286727  109.0  330.0   
regularnetglu           -0.010262  0.155439 -0.391394  0.128518  109.0  330.0   
RF                      -0.025697  0.236941 -0.478569  0.220229  109.0  330.0   
adacapnetfastselu       -0.022373  0.169191 -0.356381  0.218534  109.0  330.0   
regularnetselu          -0.023405  0.125633 -0.237110  0.159927  109.0  330.0   
adacapnetbatchstandard  -0.038370  0.170393 -0.317827  0.251934  109.0  330.0   
adacapnetstandard       -

# 5. Neural Networks with and without AdaCap

### 5.A Global performance

In [14]:
regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby('method').mean().sort_values("RMSE")[["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed"]]

RMSE        R2  Ratio_Best_RMSE_seed  \
method                                                              
adacapnetselu            0.414782  0.771019              1.139175   
regularnetglu            0.420134  0.766223              1.165654   
adacapnetglu             0.420630  0.761330              1.152071   
adacapnetresblock        0.421447  0.766774              1.182268   
regularnetstandard       0.423022  0.765300              1.220299   
adacapnetstandard        0.423317  0.763541              1.195867   
adacapnetfastselu        0.424583  0.763255              1.186416   
adacapnetbatchresblock   0.425708  0.764078              1.200271   
regularnetselu           0.426012  0.760994              1.179055   
adacapnetfast            0.428157  0.761453              1.228210   
regularnetresblock       0.428298  0.754150              1.248691   
regularnetfast           0.431209  0.761213              1.272808   
regularnetfastselu       0.439162  0.750726              1.256094   
regularnetbatchresblock  0.443639  0.741890              1.437460   
adacapnetbatchstandard   0.445570  0.747873              1.328215   
regularnetbatchstandard  0.508979  0.701647              2.343440   

                         Ratio_CAT_RMSE_seed  
method                                        
adacapnetselu                       0.983867  
regularnetglu                       0.991217  
adacapnetglu                        0.991892  
adacapnetresblock                   0.999954  
regularnetstandard                  1.003704  
adacapnetstandard                   1.007508  
adacapnetfastselu                   1.023041  
adacapnetbatchresblock              1.016476  
regularnetselu                      1.010595  
adacapnetfast                       1.024992  
regularnetresblock                  1.018829  
regularnetfast                      1.038209  
regularnetfastselu                  1.062783  
regularnetbatchresblock             1.119871  
adacapnetbatchstandard              1.096827  
regularnetbatchstandard             1.467487

## Impact of using AdaCap during training on performance 
- For each architecture each dataset and each seed, we compute the ratio of AdaCap over Regular net in terms of RMSE and time. Note that regular nets used Batch Norm and Drop out and AdaCap did not. 

### 5.B Global comparison over all datasets and all seeds

In [20]:
regression_df_formated["architecture"]=[value.replace("adacapnet","").replace("regularnet","")* bool("adacap" in value or "regular" in value) for value in regression_df_formated["method"].values]
adacap_perf = regression_df_formated[regression_df_formated["category"]== "ADACAP"].set_index("architecture")
regular_perf = regression_df_formated[regression_df_formated["category"]== "NN"].set_index("architecture")
for metric in ["RMSE", "time"]:
    adacap_perf[metric] = adacap_perf[metric].values / regular_perf[metric].values
for metric in ["RMSE", "time"]:
    adacap_perf["ratio : "+metric] = adacap_perf[metric]

### Best error reduction achieved by Adacap training over regular training for a given NN architecture when AdaCap ranked first against all methods

In [21]:
best_by_ds = regression_df_formated.groupby(["dataset_name","method","category"]).mean().reset_index().sort_values(["dataset_name","R2"]).groupby("dataset_name").last().reset_index().set_index(["dataset_name", "method"])
best_by_ds = best_by_ds[best_by_ds["category"]=="ADACAP"]
adacap_perf.groupby(["dataset_name", "method"]).mean().loc[best_by_ds.index].sort_values("ratio : RMSE")[["ratio : RMSE", "ratio : time"]]

ratio : RMSE  \
dataset_name                                       method                            
Concrete Slump Testslump_test.data_task_R_targe... adacapnetselu          0.858558   
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas... adacapnetglu           0.938864   
Electrical Grid Stability Simulated Data Data_f... adacapnetglu           0.947631   
Computer Hardwaremachine.data_task_R_target_-1     adacapnetselu          0.948213   
Concrete Slump Testslump_test.data_task_R_targe... adacapnetfastselu      0.957994   

                                                                      ratio : time  
dataset_name                                       method                           
Concrete Slump Testslump_test.data_task_R_targe... adacapnetselu          1.503810  
Yacht Hydrodynamicsyacht_hydrodynamics.data_tas... adacapnetglu           1.300586  
Electrical Grid Stability Simulated Data Data_f... adacapnetglu           2.056580  
Computer Hardwaremachine.data_task_R_target_-1     adacapnetselu          1.536705  
Concrete Slump Testslump_test.data_task_R_targe... adacapnetfastselu      1.114593

### Average Gain on datasets where AdaCap won against all methods

In [22]:
adacap_perf.groupby("architecture").mean().sort_values("RMSE").iloc[::-1][["ratio : RMSE", "ratio : time"]]

ratio : RMSE  ratio : time
architecture                             
standard           1.001047      2.541383
glu                0.997156      1.685655
fast               0.982008      2.035864
resblock           0.978925      0.872823
selu               0.972981      1.959940
fastselu           0.957026      1.356103
batchresblock      0.952732     27.049304
batchstandard      0.895591     41.420039

### Average Gain on datasets where AdaCap won against NN

In [23]:
adacap_perf[adacap_perf["RMSE"]< 1.].groupby("architecture").mean().sort_values("RMSE").iloc[::-1][["ratio : RMSE", "ratio : time"]]

ratio : RMSE  ratio : time
architecture                             
selu               0.938970      2.183972
fast               0.936266      2.145614
standard           0.926621      2.246681
glu                0.920090      1.317918
fastselu           0.907980      1.320588
resblock           0.907092      0.868544
batchresblock      0.840737     54.060377
batchstandard      0.722039     68.356523

### Average Gain on datasets where AdaCap won against all methods

In [24]:
adacap_perf[adacap_perf["Ratio_Best_RMSE_seed"]== 1.].groupby("architecture").mean().sort_values("RMSE").iloc[::-1][["ratio : RMSE", "ratio : time"]]

ratio : RMSE  ratio : time
architecture                             
fast               0.930717      1.549389
resblock           0.882422      0.863456
selu               0.874050      1.543464
glu                0.866324      1.593488
batchresblock      0.812096      1.068475
fastselu           0.809114      1.109175
standard           0.768627      2.073468

### 5.C Zoom on a specific dataset

In [25]:
dataset_index = 0 # between 0 and 25
dataset_name = adacap_perf["dataset_name"].unique()[dataset_index]
print("DATASET : "+dataset_name)
table_mean = adacap_perf.groupby(["dataset_name","architecture"]).mean().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("architecture")[['RMSE', 'time',"n","p"]]
table_std = adacap_perf.groupby(["dataset_name","architecture"]).std().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("architecture")[['RMSE', 'time',"n","p"]]
table_min = adacap_perf.groupby(["dataset_name","architecture"]).min().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("architecture")[['RMSE', 'time',"n","p"]]
table_max = adacap_perf.groupby(["dataset_name","architecture"]).max().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("architecture")[['RMSE','time',"n","p"]]
for metric in ["RMSE", "time"]:
    table_mean[metric + ":std"] = table_std[metric]
    table_mean[metric + ":min"] = table_min[metric]
    table_mean[metric + ":max"] = table_max[metric]
table_mean[[metric+suffix for metric in ["RMSE", "time"] for suffix in ["",]]+["n", "p"]]

DATASET : mirichoi0218_insuranceinsurance.csv_task_R_target_-1


RMSE      time      n      p
architecture                                   
resblock       0.948221  0.855805  109.0  330.0
selu           0.978913  1.073155  109.0  330.0
batchresblock  0.984929  1.371958  109.0  330.0
fast           0.999734  1.734509  109.0  330.0
standard       1.008511  1.807712  109.0  330.0
fastselu       1.012888  1.012483  109.0  330.0
batchstandard  1.019623  1.866549  109.0  330.0
glu            1.066076  1.076792  109.0  330.0

### 5.D Count of Top 1 performance

In [26]:
metric_id = 0 # Value between 0 and 5
metric = ["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed", "time"][metric_id]

granularity_id = 1 # Value between 0 and 1
granularity = ["method","category"][granularity_id]

metric_order = [1,-1,1,1,1][metric_id]
regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby(["dataset_name","category","method"]).mean().reset_index().sort_values(["dataset_name",metric]).iloc[::metric_order].groupby(["dataset_name"]).first()[[granularity,metric]].groupby(granularity).count().sort_values(metric).iloc[::-1]

RMSE
category      
ADACAP      14
NN          12

### 5.E Count of Top 1 performance by architecture

In [27]:
metric_id = 0 # Value between 0 and 5
metric = ["RMSE","R2","Ratio_Best_RMSE_seed", "Ratio_CAT_RMSE_seed", "time"][metric_id]

granularity_id = 0 # Value between 0 and 1
granularity = ["method","category"][granularity_id]

metric_order = [1,-1,1,1,1][metric_id]
pd.DataFrame.from_dict(data = {**{key:0 for key in regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])][granularity].unique()},**{key:value for key, value in regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby(["dataset_name","method","category"]).mean().reset_index().sort_values(["dataset_name",metric]).iloc[::metric_order].groupby(["dataset_name"]).first()[[granularity,metric]].groupby(granularity).count().sort_values(metric).iloc[::metric_order][metric].iteritems()}}, orient= "index", columns = ["#Datasets with "+ metric +" Top1 performance"]).sort_values("#Datasets with "+ metric +" Top1 performance").iloc[::-1]

#Datasets with RMSE Top1 performance
adacapnetselu                                               5
adacapnetglu                                                4
regularnetfastselu                                          4
regularnetresblock                                          3
regularnetglu                                               3
regularnetbatchresblock                                     2
adacapnetstandard                                           1
adacapnetresblock                                           1
adacapnetfastselu                                           1
adacapnetfast                                               1
adacapnetbatchresblock                                      1
adacapnetbatchstandard                                      0
regularnetstandard                                          0
regularnetselu                                              0
regularnetfast                                              0
regularnetbatchstandard                                     0

### 5.F Zoom on a specific dataset

In [28]:
dataset_index = 0 # between 0 and 25
dataset_name = regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])]["dataset_name"].unique()[dataset_index]
print("DATASET : "+dataset_name)
table_mean = regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby(["dataset_name","method","category"]).mean().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']].iloc[::-1]
table_std = regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby(["dataset_name","method","category"]).std().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']].iloc[::-1]
table_min = regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby(["dataset_name","method","category"]).min().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']].iloc[::-1]
table_max = regression_df_formated[np.isin(regression_df_formated["category"],["ADACAP","NN"])].groupby(["dataset_name","method","category"]).max().reset_index().set_index("dataset_name").loc[dataset_name].sort_values("RMSE").set_index("method")[['category', 'R2', 'RMSE', 'n', 'p', 'time']].iloc[::-1]
for metric in ["R2", "RMSE"]:
    table_mean[metric + ":std"] = table_std[metric]
    table_mean[metric + ":min"] = table_min[metric]
    table_mean[metric + ":max"] = table_max[metric]
table_mean[['category']+[metric+suffix for metric in ["RMSE","R2"] for suffix in ["", ":std",":min", ":max"]]+ [ 'n', 'p', 'time']]

DATASET : mirichoi0218_insuranceinsurance.csv_task_R_target_-1


category      RMSE  RMSE:std  RMSE:min  RMSE:max  \
method                                                                     
adacapnetglu              ADACAP  1.065250  0.055430  1.010076  1.177572   
regularnetresblock            NN  1.041671  0.124313  0.877948  1.252758   
regularnetstandard            NN  1.017816  0.118279  0.883976  1.303262   
adacapnetstandard         ADACAP  1.017145  0.082787  0.864908  1.124677   
adacapnetbatchstandard    ADACAP  1.015803  0.085081  0.864908  1.147966   
regularnetselu                NN  1.009929  0.061894  0.916555  1.112255   
adacapnetfastselu         ADACAP  1.008141  0.081821  0.884006  1.164638   
regularnetglu                 NN  1.002677  0.073788  0.933532  1.179574   
regularnetbatchresblock       NN  0.997497  0.103698  0.873633  1.205637   
regularnetbatchstandard       NN  0.997442  0.064485  0.882504  1.100568   
regularnetfastselu            NN  0.997177  0.064996  0.920170  1.148371   
regularnetfast                NN  0.993271  0.085683  0.829759  1.112328   
adacapnetfast             ADACAP  0.991107  0.086218  0.834141  1.108491   
adacapnetselu             ADACAP  0.987899  0.082149  0.819413  1.125852   
adacapnetresblock         ADACAP  0.980095  0.067166  0.813040  1.049151   
adacapnetbatchresblock    ADACAP  0.977853  0.066826  0.813040  1.049151   

                               R2    R2:std    R2:min    R2:max      n      p  \
method                                                                          
adacapnetglu            -0.137522  0.121479 -0.386675 -0.020255  109.0  330.0   
regularnetresblock      -0.098986  0.262068 -0.569403  0.229207  109.0  330.0   
regularnetstandard      -0.048541  0.261143 -0.698491  0.218586  109.0  330.0   
adacapnetstandard       -0.040752  0.164643 -0.264898  0.251934  109.0  330.0   
adacapnetbatchstandard  -0.038370  0.170393 -0.317827  0.251934  109.0  330.0   
regularnetselu          -0.023405  0.125633 -0.237110  0.159927  109.0  330.0   
adacapnetfastselu       -0.022373  0.169191 -0.356381  0.218534  109.0  330.0   
regularnetglu           -0.010262  0.155439 -0.391394  0.128518  109.0  330.0   
regularnetbatchresblock -0.004678  0.217799 -0.453561  0.236766  109.0  330.0   
regularnetbatchstandard  0.001366  0.128835 -0.211250  0.221187  109.0  330.0   
regularnetfastselu       0.001835  0.134553 -0.318755  0.153287  109.0  330.0   
regularnetfast           0.006805  0.166676 -0.237274  0.311500  109.0  330.0   
adacapnetfast            0.011017  0.167162 -0.228752  0.304209  109.0  330.0   
adacapnetselu            0.017981  0.160006 -0.267542  0.328562  109.0  330.0   
adacapnetresblock        0.035354  0.124966 -0.100717  0.338966  109.0  330.0   
adacapnetbatchresblock   0.039784  0.124371 -0.100717  0.338966  109.0  330.0   

                              time  
method                              
adacapnetglu             19.175377  
regularnetresblock        4.007936  
regularnetstandard        2.122999  
adacapnetstandard         3.820403  
adacapnetbatchstandard    2.126351  
regularnetselu           15.549521  
adacapnetfastselu        15.814718  
regularnetglu            17.805353  
regularnetbatchresblock   1.782917  
regularnetbatchstandard   1.590106  
regularnetfastselu       15.621853  
regularnetfast            0.838841  
adacapnetfast             1.436123  
adacapnetselu            16.690391  
adacapnetresblock         3.429815  
adacapnetbatchresblock    2.193559

# 6. Generate TABLE 3 

In [1]:
# Load data
import numpy as np
import pandas as pd
regression_df_formated = pd.read_csv("results/regression_benchmark.csv", index_col = 0)
regression_df_formated["R2"] = regression_df_formated["R2"].values.astype(float)
regression_df_formated["RMSE"] = np.sqrt(1. - regression_df_formated["R2"].values)

# Compute ratio over best performing method in terms of RMSE 
# See section 1.C for explanation
regression_df_formated.set_index(["dataset_name","dataset_seed"], inplace = True)
regression_df_formated["Best_RMSE_seed"] = regression_df_formated.reset_index().groupby(["dataset_name","dataset_seed"]).min()["RMSE"]
regression_df_formated.reset_index(inplace = True)
regression_df_formated["Ratio_Best_RMSE_seed"] = (regression_df_formated["RMSE"] / regression_df_formated["Best_RMSE_seed"])

# Extract best method of each category
best_of_each = {}
for category in regression_df_formated["category"].unique():
    best_of_each[category] = regression_df_formated[regression_df_formated["category"]== category].groupby("method").mean().sort_values("Ratio_Best_RMSE_seed").index.values[0]
only_best_df = regression_df_formated[np.isin(regression_df_formated["method"].values,list(best_of_each.values()))]

# For each method get list of datasets on which it outperforms all others
sort_metric = "Ratio_Best_RMSE_seed"
datasets_of_each = {}
selected_categories = only_best_df.groupby(["dataset_name","method","category"]).mean().reset_index().sort_values(["dataset_name",sort_metric]).groupby("dataset_name").first().reset_index()["category"].unique()
for category in selected_categories:
    datasets_of_each[category] = only_best_df.groupby(["dataset_name","method","category"]).mean().reset_index().sort_values(["dataset_name",sort_metric]).groupby("dataset_name").first().reset_index().set_index("category").loc[[category]]["dataset_name"].values

# Aggregate datasets based on which method outperform all others    
table_by_category = {}
for category in selected_categories:
    table_by_category[category] = only_best_df.set_index("dataset_name").loc[set(datasets_of_each[category])].reset_index().groupby(["category", "method"]).mean().reset_index()
    table_by_category[category]["dataset_best_method"] = category
    table_by_category[category].set_index(["dataset_best_method","method"], inplace = True)
    table_by_category[category]["count"] = len(set(datasets_of_each[category]))
concatenated = pd.concat(list(table_by_category.values())).reset_index()
best_methods = concatenated.sort_values(["dataset_best_method",sort_metric]).groupby("dataset_best_method").first()["method"].values
final_format = concatenated.pivot_table(values=sort_metric, index=["dataset_best_method"],
                    columns=["method"], aggfunc=np.mean)[best_methods]

# renormalize metric : divide by baseline value, transform into %
values = final_format.values.copy()
values = values / np.diag(values).reshape((-1,1))
values = values * 100
values = values - 100

# Prettify table values
table_matrix_values = np.array([[(str(np.around(col,decimals=1))+"%")*(col > 0) + 'Best' * (col <= 0) for col in cols] for cols in values])
final_format[final_format.columns] = table_matrix_values

# Count for each method on how many datasets it achieve top1 performance
final_format["# Datasets"] = [table_by_category[category]["count"].values[0] for category in final_format.index.values]

# Prettify column and index names
cat_to_method = {'GBDT':"CAT",'ADACAP':"adacapnetselu", 'NN':"regularnetglu",'GLM':"Enet",'RF':"XRF", 'SVM':"NuSVM"}
final_format.index = [cat_to_method[value] for value in final_format.index.values]
final_format = final_format[["# Datasets"]+["CAT","adacapnetselu","regularnetglu","Enet","XRF","NuSVM"]].loc[["CAT","adacapnetselu","regularnetglu","Enet","XRF","NuSVM"]]
final_format

<ipython-input-1-b844a7befb0a>:47: DeprecationWarning: In future, it will be an error for 'np.bool_' scalars to be interpreted as an index
  table_matrix_values = np.array([[(str(np.around(col,decimals=1))+"%")*(col > 0) + 'Best' * (col <= 0) for col in cols] for cols in values])


method         # Datasets     CAT adacapnetselu regularnetglu    Enet     XRF  \
CAT                     9    Best         11.6%         13.4%   56.8%   12.6%   
adacapnetselu           5  141.3%          Best         13.6%  176.6%  103.3%   
regularnetglu           4   11.5%          5.8%          Best   92.1%   31.3%   
Enet                    4    0.9%         11.6%         10.3%    Best   21.4%   
XRF                     3   12.2%         11.2%         13.2%   39.0%    Best   
NuSVM                   1    4.1%          3.1%          4.7%    6.8%   36.8%   

method          NuSVM  
CAT             21.8%  
adacapnetselu  197.9%  
regularnetglu   58.0%  
Enet            12.6%  
XRF             40.9%  
NuSVM            Best

# 7. Exotic Pricing

Code for experiment in ExoticPricing.ipynb

In [3]:
princings = pd.read_csv("outputs/exotic.csv")

pricings.sort_values("RMSE")

FileNotFoundError: [Errno 2] No such file or directory: 'outputs/exotic.csv'